datafusion/core/src/optimizer/simplify_expressions.rs

```diff
@@ -400,9 +402,9 @@ impl<'a> ConstEvaluator<'a> {
     }
 
     /// Internal helper to evaluates an Expr
-    pub(crate) fn evaluate_to_scalar(&self, expr: Expr) -> Result<ScalarValue> {
+    pub(crate) fn evaluate_to_scalar(&self, expr: &Expr) -> Result<ScalarValue> {
         if let Expr::Literal(s) = expr {
-            return Ok(s);
+            return Ok(s.clone());
```

Comment:
The `clone` is unfortunate -- I wonder if there is some way to pass it back. Like for example a signature such as  he following (aka return the `Expr` as the error if the evaluation doesn't work)?

Maybe with some debugging messages on error (via `debug!`) would be sufficent?

```rust
    pub(crate) fn evaluate_to_scalar(&self, expr: Expr) -> std::result::Result<ScalarValue, Expr> {
```

Reply:
@alamb I reverted this change and actually completely changed direction on this PR ... please see the updated PR description.